### ISSUE TITLE
Your Projects List > Add ability to preview a project

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.mozilla.org#1628]
When all you have is the project title, it's often difficult to find a project you're looking for.

Could we add an ability to preview a project? We could include a button that opens a small popup with an iframe that has the project running inside of it.

I've tried this already on the [Thimble Dashboard](https://thimble-dashboard-staging.herokuapp.com/), and it works pretty well. Could we somehow also do it for projects that haven't been published?

What if we showed your last 8 edited projects as these live thumbnails and the rest as a list (or allowed a toggle between the two).

**Comment by flukeout** [comment]
#1627 is also a pretty good solution for this.

**Comment by gideonthomas** [comment]
There is no easy way to do the iframe solution for non-published projects because of security concerns. To load the project in an iframe, we would have to host the files somewhere. We can't host unpublished project contents on Thimble due to security concerns, if we host it anywhere else, then effectively, we are publishing the project.

The only solution I can think of for unpublished projects is to have some sort of snapshotting tool run in the editor's preview iframe that occasionally snapshots the preview and writes it to the filesystem. We then use that to display the thumbnail. But this is pretty hacky and quite susceptible to break easily.

**Comment by humphd** [comment]
There are DOM screenshot libs we could try.

**Comment by jgmac1106** [comment]
Having a screenshot tool built in would be neat on many fronts.

**Comment by humphd** [comment]
https://github.com/tsayen/dom-to-image is one.

**Comment by flukeout** [comment]
Yeah, totally forgot about that solution only working well for published projects.

@humphd Once you're signed in, is there some persistent, knowable URL to access files that aren't yet published?

**Comment by humphd** [comment]
@flukeout, I don't follow.

My thought was that when you hit "publish" we can generate a screenshot for you using the DOM, and either stick it in your project's fs, or just send it to the server.  I think we should probably keep it with your project so we don't have different files locally vs. remote.

If we name it something searchable, maybe `screenshot.png` or `preview.png`, we can then do a `fs.find(...)` to locate them all in the user's filesystem and display.

**Comment by flukeout** [comment]
@humphd If you're publishing a project we can already load it into an ``iframe`` on the My Projects page and show that to the user. The snag seems to be in trying to do show a preview for projects that haven't been published yet.

We could do the screenshot generation at some other point in the editing process though. I used a solution like this in the past and it worked pretty well, the only issue was that you couldn't save any image with an external URL if CORS stuff wasn't set up correctly on wherever that file was hosted.

I used https://html2canvas.hertzen.com/ 

Anyway, worth taking a crack at.

The reason I was asking about some way to reference non-published files is so that we could use an iframe-based solution to show an authenticated user a preview of their stuff.

Another Q - can we do 'hidden' files in the brackets filetree? Might be nice to hide the auto-generated thumbnails from view if we decide to go with that.
